addons/portal/controllers/portal.py

```diff
@@ -167,6 +186,100 @@ def _prepare_portal_overview_values(self):
             }]
         }
 
+    def _is_own_partner_address(self, partner, address_id):
+        """Checks whether address_id is the id of a res.partner representing:
+            - the main address of partner (partner.id)
+            - an invoice address from partner.child_ids
+            - a delivery address from partner.child_ids
+        """
+        # if address_id is partner's main address
+        if partner.id == address_id:
+            return True
+        # if address_id partner's secondary address
+        return address_id in partner.child_ids.filtered(lambda partner: partner.type in ['invoice', 'delivery']).ids
+
+    def _render_portal_my_address(self, values):
+        response = request.render("portal.portal_my_address", values)
+        response.headers['X-Frame-Options'] = 'DENY'
+        return response
+
+    def _prepare_address_form_values(self, address_id=None):
+        values = self._prepare_portal_layout_values()
+
+        partner = request.env.user.partner_id
+        address = request.env['res.partner'].with_context(show_address=1).browse(address_id) if address_id else False
+
+        if address:
+            fields = {**self.MANDATORY_PARTNER_FIELDS, **self.OPTIONAL_PARTNER_FIELDS}
+            # add address data
+            values.update({value: getattr(address, key) for key, value in fields.items()})
+
+        countries = request.env['res.country'].sudo().search([])
+        states = request.env['res.country.state'].sudo().search([])
+
+        values.update({
+            'partner': partner,
+            'address': address,
+            'countries': countries,
+            'states': states,
+            'has_check_vat': hasattr(request.env['res.partner'], 'check_vat'),
+            'page_name': 'my_details',
+        })
+        return values
+
+    @route(['/my/address/create', '/my/address/<int:address_id>/update'], type='http', auth='user', website=True, methods=['GET', 'POST'])
+    def address_write(self, address_id=None, redirect=None, **post):
```

Comment:
I would split it in two controllers, one for create and one for write. Even if it implies a bit more code (maybe we could extract some code into tool sub methods), would be clearer I think.